File: javascript/20_Valid_Parentheses.js
var isValid = function(s) {
    stack = [];
    map ={
        ')': '(',
        ']': '[',
        '}': '{'
    };

    for (c of s) {
        console.log(stack);
        if (c in map) {
            if (stack && stack[stack.length - 1] === map[c]) {
                stack.pop();
            } 
            else return false;
        } 
        else stack.push(c);
    }
    return !stack.length ? true : false;
};

output1 = isValid("()");
output2 = isValid("()[]{}");
output3 = isValid("(]");
console.log(output1, output2, output3);

File: python/string/20._Valid_Parentheses.py
'''
Key:
1. lost as many cases as possible 
2. in python: stack(list), hashmap(dict)
'''

class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        map = {')': '(', ']': '[', '}': '{'}

        for c in s:
            if c in map:
                if stack and stack[-1] == map[c]:  # if stack is not empty, otherwise index may be out of range
                    stack.pop()
                else:
                    return False
            else:
                stack.append(c)

        return True if not stack else False


output1 = Solution().isValid("()")
output2 = Solution().isValid("()[]{}")
output3 = Solution().isValid("(]")
print(output1, output2, output3)